public/js/components/ObjectInspector.js

```diff
@@ -0,0 +1,169 @@
+"use strict";
+
+const React = require("react");
+const classnames = require("classnames");
+const ManagedTree = React.createFactory(require("./util/ManagedTree"));
+const Arrow = React.createFactory(require("./util/Arrow"));
+const Rep = require("./Rep");
+const { DOM: dom, PropTypes } = React;
+
+// This implements a component that renders an interactive inspector
+// for looking at JavaScript objects. It expects descriptions of
+// objects from the protocol, and will dynamically fetch child
+// properties as objects are expanded.
+//
+// If you want to inspect a single object, pass the name and the
+// protocol descriptor of it:
+//
+//  ObjectInspector({
+//    name: "foo",
+//    desc: { writable: true, ..., { value: { actor: "1", ... }}},
+//    ...
+//  })
+//
+// If you want multiple top-level objects (like scopes), you can pass
+// an array of manually constructed nodes as `roots`:
+//
+//  ObjectInspector({
+//    roots: [{ name: ... }, ...],
+//    ...
+//  });
+
+// There are 3 types of nodes: a simple node with a children array, an
+// object that has properties that should be children when they are
+// fetched, and a primitive value that should be displayed with no
+// children.
+
+function nodeHasChildren(item) {
+  // Do not use `Array.isArray` because it's slower and we do not need
+  // to support multiple globals here.
+  return item.contents instanceof Array;
+}
+
+function nodeHasProperties(item) {
+  return !nodeHasChildren(item) && item.contents.value.type === "object";
+}
+
+function nodeIsPrimitive(item) {
+  return !nodeHasChildren(item) && !nodeHasProperties(item);
+}
+
+function createNode(name, path, contents) {
+  // The path is important to uniquely identify the item in the entire
+  // tree. This helps debugging & optimizes React's rendering of large
+  // lists. The path will be separated by property name,
+  // i.e. `{ foo: { bar: { baz: 5 }}}` will have a path of `foo/bar/baz`
+  // for the inner object.
+  return { name, path, contents };
+}
+
+const ObjectInspector = React.createClass({
+  propTypes: {
+    name: PropTypes.string,
+    desc: PropTypes.object,
+    roots: PropTypes.array,
+    getObjectProperties: PropTypes.func.isRequired,
+    loadObjectProperties: PropTypes.func.isRequired
+  },
+
+  displayName: "ObjectInspector",
+
+  getInitialState() {
+    // Cache of dynamically built nodes. We shouldn't need to clear
+    // this out ever, since we don't ever "switch out" the object
+    // being inspected.
+    this.actorCache = {};
+    return {};
+  },
+
+  getChildren(item) {
+    const { getObjectProperties } = this.props;
+    const obj = item.contents;
+
+    // Nodes can either have children already, or be an object with
+    // properties that we need to go and fetch.
+    if (nodeHasChildren(item)) {
+      return item.contents;
+    } else if (nodeHasProperties(item)) {
+      const actor = obj.value.actor;
```

Comment:
Yes they are specific. We haven't made the scopes generic yet. It should be clearer how to do that now, but I'm not sure. We'll do it in the next few weeks probably.